File: react-native/src/config/dynamicConfig.ts
import { internalConsole } from '../utils/internalConsole';
import { ScreenshotMaskLevel } from './screenshotMaskLevel';

export interface IDynamicConfig {
  /** Maximum number of events and spans in a batch. Defaults to 10000. */
  maxEventsInBatch: number;

  /** Duration of session timeline collected with an ANR, in seconds. Defaults to 300. */
  anrTimelineDurationSeconds: number;

  /** Duration of session timeline collected with a bug report, in seconds. Defaults to 300. */
  bugReportTimelineDurationSeconds: number;

  /** Sampling rate for traces. Defaults to 100 */
  traceSamplingRate: number;

  /** Sampling rate for sessions that should track journey events. Defaults to 100 */
  journeySamplingRate: number;

  /** Screenshot masking level */
  screenshotMaskLevel: ScreenshotMaskLevel;

  /** Whether the SDK automatically collects logs from the platform's logging APIs. Manually tracked logs are always collected. Defaults to false. */
  logAutocollectEnabled: boolean;

  /** Minimum severity number of logs to collect. Logs below this number are dropped. Defaults to 16 (warning). */
  logMinSeverity: number;

  /** Regex patterns matched against the log body. A log whose body matches any of the patterns is dropped. Defaults to empty list. */
  logIgnorePatterns: string[];

  /** Interval in seconds to collect CPU usage. Defaults to 5. */
  cpuUsageInterval: number;

  /** Interval in seconds to collect memory usage. Defaults to 5. */
  memoryUsageInterval: number;

  /** Whether to take a screenshot on ANR. Defaults to true. */
  anrTakeScreenshot: boolean;

  /** Sampling rate for launch metrics. Defaults to 100 */
  launchSamplingRate: number;

  /** Whether to take a layout snapshot on gesture click. Defaults to true. */
  gestureClickTakeSnapshot: boolean;

  /** URLs for which HTTP events should be disabled. */
  httpDisableEventForUrls: string[];

  /** URLs for which HTTP requests should be tracked. */
  httpTrackRequestForUrls: string[];

  /** URLs for which HTTP responses should be tracked. */
  httpTrackResponseForUrls: string[];

  /** HTTP headers that should never be collected. */
  httpBlockedHeaders: string[];
}

export class DynamicConfig implements IDynamicConfig {
  maxEventsInBatch: number;
  anrTimelineDurationSeconds: number;
  bugReportTimelineDurationSeconds: number;
  traceSamplingRate: number;
  journeySamplingRate: number;
  screenshotMaskLevel: ScreenshotMaskLevel;
  logAutocollectEnabled: boolean;
  logMinSeverity: number;
  logIgnorePatterns: string[];
  cpuUsageInterval: number;
  memoryUsageInterval: number;
  anrTakeScreenshot: boolean;
  launchSamplingRate: number;
  gestureClickTakeSnapshot: boolean;
  httpDisableEventForUrls: string[];
  httpTrackRequestForUrls: string[];
  httpTrackResponseForUrls: string[];
  httpBlockedHeaders: string[];

  constructor(values: IDynamicConfig) {
    this.maxEventsInBatch = values.maxEventsInBatch;
    this.anrTimelineDurationSeconds = values.anrTimelineDurationSeconds;
    this.bugReportTimelineDurationSeconds =
      values.bugReportTimelineDurationSeconds;
    this.traceSamplingRate = values.traceSamplingRate;
    this.journeySamplingRate = values.journeySamplingRate;
    this.screenshotMaskLevel = values.screenshotMaskLevel;
    this.logAutocollectEnabled = values.logAutocollectEnabled;
    this.logMinSeverity = values.logMinSeverity;
    this.logIgnorePatterns = values.logIgnorePatterns;
    this.cpuUsageInterval = values.cpuUsageInterval;
    this.memoryUsageInterval = values.memoryUsageInterval;
    this.anrTakeScreenshot = values.anrTakeScreenshot;
    this.launchSamplingRate = values.launchSamplingRate;
    this.gestureClickTakeSnapshot = values.gestureClickTakeSnapshot;
    this.httpDisableEventForUrls = values.httpDisableEventForUrls;
    this.httpTrackRequestForUrls = values.httpTrackRequestForUrls;
    this.httpTrackResponseForUrls = values.httpTrackResponseForUrls;
    this.httpBlockedHeaders = values.httpBlockedHeaders;
  }

  static default(): DynamicConfig {
    return new DynamicConfig({
      maxEventsInBatch: 10_000,
      anrTimelineDurationSeconds: 300,
      bugReportTimelineDurationSeconds: 300,
      traceSamplingRate: 100,
      journeySamplingRate: 100,
      screenshotMaskLevel: ScreenshotMaskLevel.allTextAndMedia,
      logAutocollectEnabled: false,
      logMinSeverity: 16,
      logIgnorePatterns: [],
      cpuUsageInterval: 5,
      memoryUsageInterval: 5,
      anrTakeScreenshot: true,
      launchSamplingRate: 100,
      gestureClickTakeSnapshot: true,
      httpDisableEventForUrls: [],
      httpTrackRequestForUrls: [],
      httpTrackResponseForUrls: [],
      httpBlockedHeaders: [
        'Authorization',
        'Cookie',
        'Set-Cookie',
        'Proxy-Authorization',
        'WWW-Authenticate',
        'X-Api-Key',
      ],
    });
  }

  /**
   * Creates DynamicConfig from native object (snake_case keys)
   */
  static fromNative(obj: any): DynamicConfig {
    if (!obj || typeof obj !== 'object') {
      return DynamicConfig.default();
    }

    const defaults = DynamicConfig.default();

    return new DynamicConfig({
      maxEventsInBatch: numberOrDefault(
        obj,
        'max_events_in_batch',
        defaults.maxEventsInBatch
      ),
      anrTimelineDurationSeconds: numberOrDefault(
        obj,
        'anr_timeline_duration',
        defaults.anrTimelineDurationSeconds
      ),
      bugReportTimelineDurationSeconds: numberOrDefault(
        obj,
        'bug_report_timeline_duration',
        defaults.bugReportTimelineDurationSeconds
      ),
      traceSamplingRate: numberOrDefault(
        obj,
        'trace_sampling_rate',
        defaults.traceSamplingRate
      ),
      journeySamplingRate: numberOrDefault(
        obj,
        'journey_sampling_rate',
        defaults.journeySamplingRate
      ),
      screenshotMaskLevel: screenshotMaskLevelOrDefault(
        obj,
        'screenshot_mask_level',
        defaults.screenshotMaskLevel
      ),
      logAutocollectEnabled: booleanOrDefault(
        obj,
        'log_autocollect_enabled',
        defaults.logAutocollectEnabled
      ),
      logMinSeverity: numberOrDefault(
        obj,
        'log_min_severity',
        defaults.logMinSeverity
      ),
      logIgnorePatterns: stringArrayOrDefault(
        obj,
        'log_ignore_patterns',
        defaults.logIgnorePatterns
      ),
      cpuUsageInterval: numberOrDefault(
        obj,
        'cpu_usage_interval',
        defaults.cpuUsageInterval
      ),
      memoryUsageInterval: numberOrDefault(
        obj,
        'memory_usage_interval',
        defaults.memoryUsageInterval
      ),
      anrTakeScreenshot: booleanOrDefault(
        obj,
        'anr_take_screenshot',
        defaults.anrTakeScreenshot
      ),
      launchSamplingRate: numberOrDefault(
        obj,
        'launch_sampling_rate',
        defaults.launchSamplingRate
      ),
      gestureClickTakeSnapshot: booleanOrDefault(
        obj,
        'gesture_click_take_snapshot',
        defaults.gestureClickTakeSnapshot
      ),
      httpDisableEventForUrls: stringArrayOrDefault(
        obj,
        'http_disable_event_for_urls',
        defaults.httpDisableEventForUrls
      ),
      httpTrackRequestForUrls: stringArrayOrDefault(
        obj,
        'http_track_request_for_urls',
        defaults.httpTrackRequestForUrls
      ),
      httpTrackResponseForUrls: stringArrayOrDefault(
        obj,
        'http_track_response_for_urls',
        defaults.httpTrackResponseForUrls
      ),
      httpBlockedHeaders: stringArrayOrDefault(
        obj,
        'http_blocked_headers',
        defaults.httpBlockedHeaders
      ),
    });
  }
}

function warnInvalidField(key: string, expected: string): void {
  internalConsole.warn(
    `[Measure] Dynamic config field '${key}' expected ${expected}, using default.`
  );
}

function numberOrDefault(obj: any, key: string, fallback: number): number {
  const value = obj[key];
  if (value === undefined || value === null) {
    return fallback;
  }
  if (typeof value !== 'number' || !Number.isFinite(value)) {
    warnInvalidField(key, 'a number');
    return fallback;
  }
  return value;
}

function booleanOrDefault(obj: any, key: string, fallback: boolean): boolean {
  const value = obj[key];
  if (value === undefined || value === null) {
    return fallback;
  }
  if (typeof value !== 'boolean') {
    warnInvalidField(key, 'a boolean');
    return fallback;
  }
  return value;
}

function stringArrayOrDefault(
  obj: any,
  key: string,
  fallback: string[]
): string[] {
  const value = obj[key];
  if (value === undefined || value === null) {
    return fallback;
  }
  if (
    !Array.isArray(value) ||
    !value.every((item) => typeof item === 'string')
  ) {
    warnInvalidField(key, 'a string array');
    return fallback;
  }
  return value;
}

function screenshotMaskLevelOrDefault(
  obj: any,
  key: string,
  fallback: ScreenshotMaskLevel
): ScreenshotMaskLevel {
  const value = obj[key];
  if (value === undefined || value === null) {
    return fallback;
  }
  if (
    typeof value !== 'string' ||
    !Object.values(ScreenshotMaskLevel).includes(value as ScreenshotMaskLevel)
  ) {
    warnInvalidField(key, 'a valid ScreenshotMaskLevel');
    return fallback;
  }
  return value as ScreenshotMaskLevel;
}

File: android/measure-android/measure/src/main/java/sh/measure/android/config/DynamicConfig.kt
package sh.measure.android.config

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

internal interface IDynamicConfig {
    /**
     * Maximum number of events and spans in a batch. Defaults to 10000.
     */
    val maxEventsInBatch: Int

    /**
     * Duration of session timeline collected with an error, in seconds. Defaults to 300 seconds.
     */
    val errorReplayDurationSeconds: Int

    /**
     * Duration of session timeline collected with an ANR, in seconds. Defaults to 300 seconds.
     */
    val anrTimelineDurationSeconds: Int

    /**
     * Duration of session timeline collected with a bug report, in seconds. Defaults to 300 seconds.
     */
    val bugReportTimelineDurationSeconds: Int

    /**
     * Sampling rate for traces. Defaults to 100%, which means all traces will be sampled.
     */
    val traceSamplingRate: Float

    /**
     * Sampling rate in percentage for sessions that should track journey events. Defaults
     * to 100%, which means all sessions will track journey events.
     */
    val journeySamplingRate: Float

    /**
     * The screenshot mask level to be applied to screenshots. Defaults
     * to [ScreenshotMaskLevel.AllTextAndMedia].
     */
    val screenshotMaskLevel: ScreenshotMaskLevel

    /**
     * Whether the SDK automatically collects logs from the platform's logging APIs.
     * Manually tracked logs are always collected. Defaults to false.
     */
    val logAutocollectEnabled: Boolean

    /**
     * Minimum severity number of logs to collect. Logs below this number are dropped.
     * Defaults to 16 (warning).
     */
    val logMinSeverity: Int

    /**
     * Regex patterns matched against the log body. A log whose body matches any of the
     * patterns is dropped before it is tracked. Defaults to empty list.
     */
    val logIgnorePatterns: List<String>

    /**
     * Interval in seconds to collect CPU usage. Defaults to 5 seconds.
     */
    val cpuUsageInterval: Long

    /**
     * Interval in seconds to collect memory usage. Defaults to 5 seconds.
     */
    val memoryUsageInterval: Long

    /**
     * Whether to take a screenshot when a fatal error occurs. Defaults to true.
     */
    val errorFatalScreenshotEnabled: Boolean

    /**
     * Whether to collect a session replay with fatal errors, the ones that terminated
     * the app. Defaults to true.
     */
    val errorFatalReplayEnabled: Boolean

    /**
     * Whether to collect a session replay with unhandled errors, the ones that were not
     * caught but did not terminate the app. Defaults to false.
     */
    val errorUnhandledReplayEnabled: Boolean

    /**
     * Whether to collect a session replay with handled errors, the ones reported by the
     * app. Defaults to false.
     */
    val errorHandledReplayEnabled: Boolean

    /**
     * Sampling rate for fatal errors. Defaults to 100.
     */
    val errorFatalSamplingRate: Float

    /**
     * Sampling rate for unhandled errors. Defaults to 100.
     */
    val errorUnhandledSamplingRate: Float

    /**
     * Sampling rate for handled errors. Defaults to 0.
     */
    val errorHandledSamplingRate: Float

    /**
     * Whether to take a screenshot when an ANR occurs. Defaults to true.
     */
    val anrTakeScreenshot: Boolean

    /**
     * Sampling rate for launch metrics in percentage. Defaults to 100%, i.e., all launches.
     */
    val launchSamplingRate: Float

    /**
     * Whether to take a layout snapshot when a gesture click occurs. Defaults to true.
     */
    val gestureClickTakeSnapshot: Boolean

    /**
     * Sampling rate for HTTP events. Defaults to 100%, which means
     * all HTTP events will be sampled.
     */
    val httpSamplingRate: Float

    /**
     * List of URLs to disable sending events for. Defaults to empty list.
     *
     * The URLs can use wildcard patterns.
     */
    val httpDisableEventForUrls: List<String>

    /**
     * List of URLs to track requests for. Defaults to empty list.
     *
     * The URLs can use wildcard patterns.
     */
    val httpTrackRequestForUrls: List<String>

    /**
     * List of URLs to track responses for. Defaults to empty list.
     *
     * The URLs can use wildcard patterns.
     */
    val httpTrackResponseForUrls: List<String>

    /**
     * List of HTTP headers to not collect with the `http` event for both request and response.
     *
     * The following headers are always excluded:
     * * Authorization
     * * Cookie
     * * Set-Cookie
     * * Proxy-Authorization
     * * WWW-Authenticate
     * * X-Api-Key
     */
    val httpBlockedHeaders: List<String>

    /**
     * Sampling rate in percentage for profile events. Defaults to 100%, i.e. every profile that
     * survives the system's own throttling is collected. The OS already rate limits these, so this
     * is an additional, remotely tunable cap on top of that.
     */
    val profileSamplingRate: Float
}

@Serializable
internal data class DynamicConfig(
    @SerialName("max_events_in_batch")
    override val maxEventsInBatch: Int = 10000,

    @SerialName("error_replay_duration")
    val errorReplayDuration: Int? = null,

    /**
     * Renamed to `error_replay_duration`. Only
     * read when the the new key is absent.
     */
    @Deprecated("Use errorReplayDurationSeconds instead.")
    @SerialName("crash_timeline_duration")
    val crashTimelineDuration: Int? = null,

    @SerialName("anr_timeline_duration")
    override val anrTimelineDurationSeconds: Int = 300,

    @SerialName("bug_report_timeline_duration")
    override val bugReportTimelineDurationSeconds: Int = 300,

    @SerialName("trace_sampling_rate")
    override val traceSamplingRate: Float = 100f,

    @SerialName("journey_sampling_rate")
    override val journeySamplingRate: Float = 100f,

    @SerialName("screenshot_mask_level")
    override val screenshotMaskLevel: ScreenshotMaskLevel =
        ScreenshotMaskLevel.AllTextAndMedia,

    @SerialName("log_autocollect_enabled")
    override val logAutocollectEnabled: Boolean = false,

    @SerialName("log_min_severity")
    override val logMinSeverity: Int = 16,

    @SerialName("log_ignore_patterns")
    override val logIgnorePatterns: List<String> = emptyList(),

    @SerialName("cpu_usage_interval")
    override val cpuUsageInterval: Long = 5,

    @SerialName("memory_usage_interval")
    override val memoryUsageInterval: Long = 5,

    @SerialName("error_fatal_take_screenshot")
    val errorFatalTakeScreenshot: Boolean? = null,

    /**
     * Renamed to `error_fatal_take_screenshot`. Only
     * read when the the new key is absent.
     */
    @Deprecated("Use errorFatalScreenshotEnabled instead.")
    @SerialName("crash_take_screenshot")
    val crashTakeScreenshot: Boolean? = null,

    @SerialName("error_fatal_replay_enabled")
    override val errorFatalReplayEnabled: Boolean = true,

    @SerialName("error_unhandled_replay_enabled")
    override val errorUnhandledReplayEnabled: Boolean = false,

    @SerialName("error_handled_replay_enabled")
    override val errorHandledReplayEnabled: Boolean = false,

    @SerialName("error_fatal_sampling_rate")
    override val errorFatalSamplingRate: Float = 100f,

    @SerialName("error_unhandled_sampling_rate")
    override val errorUnhandledSamplingRate: Float = 100f,

    @SerialName("error_handled_sampling_rate")
    override val errorHandledSamplingRate: Float = 0f,

    @SerialName("anr_take_screenshot")
    override val anrTakeScreenshot: Boolean = true,

    @SerialName("launch_sampling_rate")
    override val launchSamplingRate: Float = 100f,

    @SerialName("gesture_click_take_snapshot")
    override val gestureClickTakeSnapshot: Boolean = true,

    @SerialName("http_sampling_rate")
    override val httpSamplingRate: Float = 100f,

    @SerialName("http_disable_event_for_urls")
    override val httpDisableEventForUrls: List<String> = emptyList(),

    @SerialName("http_track_request_for_urls")
    override val httpTrackRequestForUrls: List<String> = emptyList(),

    @SerialName("http_track_response_for_urls")
    override val httpTrackResponseForUrls: List<String> = emptyList(),

    @SerialName("http_blocked_headers")
    override val httpBlockedHeaders: List<String> = listOf(
        "Authorization",
        "Cookie",
        "Set-Cookie",
        "Proxy-Authorization",
        "WWW-Authenticate",
        "X-Api-Key",
    ),

    @SerialName("profile_sampling_rate")
    override val profileSamplingRate: Float = 100f,
) : IDynamicConfig {
    @Suppress("DEPRECATION")
    override val errorReplayDurationSeconds: Int
        get() = errorReplayDuration ?: crashTimelineDuration ?: 300

    @Suppress("DEPRECATION")
    override val errorFatalScreenshotEnabled: Boolean
        get() = errorFatalTakeScreenshot ?: crashTakeScreenshot ?: true
}
